gateway/ln-gateway/src/lib.rs

```diff
@@ -860,6 +492,140 @@ impl Gateway {
         ))
     }
 
+    async fn handle_pay_invoice_msg(&self, payload: PayInvoicePayload) -> Result<Preimage> {
+        if let GatewayState::Running {
+            lnrpc: _,
+            lightning_public_key: _,
+            lightning_alias: _,
+        } = self.state.read().await.clone()
+        {
+            let PayInvoicePayload {
+                federation_id,
+                contract_id,
+            } = payload;
+
+            let client = self.select_client(federation_id).await?;
+            let operation_id = client.gateway_pay_bolt11_invoice(contract_id).await?;
+            let mut updates = client
+                .gateway_subscribe_ln_pay(operation_id)
+                .await?
+                .into_stream();
+
+            while let Some(update) = updates.next().await {
+                match update {
+                    GatewayExtPayStates::Success {
+                        preimage,
+                        outpoint: _,
+                    } => return Ok(preimage),
+                    GatewayExtPayStates::Fail {
+                        error,
+                        error_message,
+                    } => {
+                        error!(error_message);
+                        return Err(GatewayError::OutgoingPaymentError(Box::new(error)));
+                    }
+                    GatewayExtPayStates::Canceled { error } => {
+                        return Err(GatewayError::OutgoingPaymentError(Box::new(error)));
+                    }
+                    _ => {}
+                };
+            }
+
+            return Err(GatewayError::UnexpectedState(
+                "Ran out of state updates while paying invoice".to_string(),
+            ));
+        }
+
+        Err(GatewayError::Disconnected)
+    }
+
+    async fn handle_connect_federation(
+        &mut self,
+        payload: ConnectFedPayload,
+    ) -> Result<FederationInfo> {
+        if let GatewayState::Running {
+            lnrpc,
+            lightning_public_key,
+            lightning_alias: _,
+        } = self.state.read().await.clone()
+        {
+            let invite_code = InviteCode::from_str(&payload.invite_code).map_err(|e| {
+                GatewayError::InvalidMetadata(format!("Invalid federation member string {e}"))
```

Comment:
Use the debug format so the stack trace will be included
```suggestion
                GatewayError::InvalidMetadata(format!("Invalid federation member string {e:?}"))
```